src/dsp_tools/utils/xmlupload/list_client.py

```diff
@@ -0,0 +1,120 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Any, Iterable, Protocol
+from urllib.parse import quote_plus
+
+import requests
+
+from dsp_tools.models.exceptions import BaseError
+from dsp_tools.utils.create_logger import get_logger
+
+# pylint: disable=too-few-public-methods
+
+logger = get_logger(__name__)
+
+
+@dataclass(frozen=True)
+class ListNode:
+    """Information about a list node."""
+
+    node_iri: str
+    node_name: str
+
+
+@dataclass(frozen=True)
+class List:
+    """Information about a list."""
+
+    root_iri: str
+    list_name: str
+    children: list[ListNode]
+
+
+@dataclass(frozen=True)
+class ProjectLists:
+    """Information about the lists of a project."""
+
+    lists: list[List]
+
+
+class ListClient(Protocol):
+    """Interface (protocol) for list-related requests to the DSP-API."""
+
+    def get_list_node_id_to_iri_lookup(self) -> dict[str, str]:
+        """Get a lookup of list node names to IRIs."""
+
+
+@dataclass
+class ListClientLive:
+    """Client handling list-related requests to the DSP-API."""
+
+    server: str
+    project_iri: str
+    list_info: ProjectLists | None = field(init=False, default=None)
+
+    def get_list_node_id_to_iri_lookup(self) -> dict[str, str]:
+        """
+        Get a mapping of list node IDs to their respective IRIs.
+        A list node ID is structured as follows:
+        <list name>:<node name> where the list name is the node name of the root node.
+
+        Returns:
+            The mapping of list node IDs to IRIs.
+        """
+        if not self.list_info:
+            self.list_info = _get_list_info_from_server(self.server, self.project_iri)
+        return dict(_get_node_tuples(self.list_info.lists))
+
+
+def _get_node_tuples(lists: list[List]) -> Iterable[tuple[str, str]]:
```

Comment:
Doesn't actually make any difference in this context

Reply:
But I still wonder what the correct annotation is. Every time I learn something new, I prefer putting it correct in dsp-tools, so that I have a reference for later. And the annotations of iterator/iterable/generator are something I'm still not 100% sure about yet.

Reply:
Both are correct :)  
Technically, here `Generator` would be most precise, but it would lead to a cluttered type annotation because of parts of the Generator that are not even used here; so that one I would rule out.  
As to Iterator vs. Iterable: I'm not 100% which one is more precise, probably Iterator, but both are correct because every Iterator is also an Iterable. And from a caller perspective, Iterable is what you probably want to see, because that just tells you "something that can be iterated over".